libraries/GCS_MAVLink/GCS.cpp

```diff
@@ -79,3 +79,72 @@ bool GCS::install_alternative_protocol(mavlink_channel_t c, GCS_MAVLINK::protoco
 }
 
 #undef FOR_EACH_ACTIVE_CHANNEL
+
+void GCS::update_sensor_status_flags()
+{
+    control_sensors_present = 0;
+    control_sensors_enabled = 0;
+    control_sensors_health = 0;
+
+    AP_AHRS &ahrs = AP::ahrs();
+    const AP_InertialSensor &ins = AP::ins();
+
+    control_sensors_present |= MAV_SYS_STATUS_AHRS;
+    control_sensors_enabled |= MAV_SYS_STATUS_AHRS;
+    if (!ahrs.initialised() || ahrs.healthy()) {
+        if (!ahrs.have_inertial_nav() || ins.accel_calibrated_ok_all()) {
+            control_sensors_health |= MAV_SYS_STATUS_AHRS;
+        }
+    }
+
+    const AP_Baro &barometer = AP::baro();
+    control_sensors_present |= MAV_SYS_STATUS_SENSOR_ABSOLUTE_PRESSURE;
+    control_sensors_enabled |= MAV_SYS_STATUS_SENSOR_ABSOLUTE_PRESSURE;
+    if (barometer.all_healthy()) {
+        control_sensors_health |= MAV_SYS_STATUS_SENSOR_ABSOLUTE_PRESSURE;
+    }
+
+    const AP_BattMonitor &battery = AP::battery();
+    control_sensors_present |= MAV_SYS_STATUS_SENSOR_BATTERY;
+    if (battery.num_instances() > 0) {
+        control_sensors_enabled |= MAV_SYS_STATUS_SENSOR_BATTERY;
+    }
+    if (battery.healthy() && !battery.has_failsafed()) {
+        control_sensors_health |= MAV_SYS_STATUS_SENSOR_BATTERY;
+    }
+
+    control_sensors_present |= MAV_SYS_STATUS_SENSOR_3D_GYRO;
+    if (!ins.calibrating()) {
+        control_sensors_enabled |= MAV_SYS_STATUS_SENSOR_3D_GYRO;
+        if (ins.get_gyro_health_all() && ins.gyro_calibrated_ok_all()) {
+            control_sensors_health |= MAV_SYS_STATUS_SENSOR_3D_GYRO;
+        }
+    }
+    control_sensors_present |= MAV_SYS_STATUS_SENSOR_3D_ACCEL;
+    if (!ins.calibrating()) {
+        control_sensors_enabled |= MAV_SYS_STATUS_SENSOR_3D_ACCEL;
+        if (ins.get_accel_health_all()) {
+            control_sensors_health |= MAV_SYS_STATUS_SENSOR_3D_ACCEL;
+        }
+    }
+
+    const AP_Logger &logger = AP::logger();
+    if (logger.logging_present()) {  // primary logging only (usually File)
+        control_sensors_present |= MAV_SYS_STATUS_LOGGING;
+    }
+    if (logger.logging_enabled()) {
+        control_sensors_enabled |= MAV_SYS_STATUS_LOGGING;
+    }
+    if (!logger.logging_failed()) {
+        control_sensors_health |= MAV_SYS_STATUS_LOGGING;
+    }
+
+    // set motors outputs as enabled if safety switch is not disarmed (i.e. either NONE or ARMED)
+    control_sensors_present |= MAV_SYS_STATUS_SENSOR_MOTOR_OUTPUTS;
+    if (hal.util->safety_switch_state() != AP_HAL::Util::SAFETY_DISARMED) {
+        control_sensors_enabled |= MAV_SYS_STATUS_SENSOR_MOTOR_OUTPUTS;
+        control_sensors_health |= MAV_SYS_STATUS_SENSOR_MOTOR_OUTPUTS;
```

Comment:
This should always be healthy. Just because you haven't pressed the switch to enable motors doesn't mean they are unhealthy.

I'm pushing this as this is a behavior change (on plane at least), and gets in the way for ESC's with telemetry reporting health appropriately in the future.

Reply:
Will change.  For Copter we can already add something in here for our existing motor-failure-detection code.

Reply:
Done